File: my-folder/0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> mp;
        int n = s.length();
        int res = 0;

        for(int i = 0, j = 0; j < n; j++) {
            if(mp.find(s[j]) != mp.end()) {
                i = max(mp[s[j]]+1, i);
            }
            res = max(res, j-i+1);
            mp[s[j]] = j;
        }

        return res;
    }
};

File: my-folder/0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> chars = new HashMap();
        int l = 0, r = 0;
        int res = 0;
        while(r < s.length()) {
            char c = s.charAt(r);
            chars.put(c, chars.getOrDefault(c,0)+1);
            while(chars.get(c) > 1) {
                char left = s.charAt(l);
                chars.put(left, chars.get(left) - 1);
                l++;
            }

            res = Math.max(res, r-l+1);
            r++;
        }

        return res;
    }
}
